include/deal.II/base/timer_tree.h

```diff
@@ -0,0 +1,423 @@
+#include <deal.II/base/timer.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A class for wall-time measurements enabling a modular coupling of modules in
+ * a minimally invasive way.
+ */
+class TimerTree
+{
+public:
+  /**
+   * Constructor.
+   */
+  TimerTree()
+    : id("")
+  {}
+
+  /**
+   * Clears the content of this tree. Sub trees inserted
+   * into this tree via pointers to external trees are not touched.
+   */
+  void
+  clear()
+  {
+    this->id = "";
+    data     = nullptr;
+    sub_trees.clear();
+  }
+
+  /**
+   * Inserts a measured @p wall_time into the tree, by
+   * either creating a new entry in the tree if this function is called
+   * the first time with the ID @p ids, or by adding the @p wall_time to an
+   * entry already existing in the tree.
+   */
+  void
+  insert(std::vector<std::string> const ids, double const wall_time)
+  {
+    AssertThrow(ids.size() > 0, ExcMessage("empty name."));
+
+    if (this->id == "") // the tree is currently empty
+      {
+        AssertThrow(sub_trees.empty(),
+                    ExcMessage("invalid state found. aborting."));
+
+        this->id = ids[0];
+
+        if (ids.size() == 1) // leaves of tree reached, insert the data
+          {
+            data.reset(new Data());
+            data->wall_time += wall_time;
+
+            return;
+          }
+        else // go deeper
+          {
+            std::vector<std::string> remaining_id = erase_first(ids);
+
+            std::shared_ptr<TimerTree> new_tree;
+            new_tree.reset(new TimerTree());
+            new_tree->insert(remaining_id, wall_time);
+            sub_trees.push_back(new_tree);
+          }
+      }
+    else if (this->id == ids[0]) // the tree already has some entries
+      {
+        if (ids.size() == 1) // leaves of tree reached, insert the data
+          {
+            if (data.get() == nullptr)
+              data.reset(new Data());
+
+            data->wall_time += wall_time;
+
+            return;
+          }
+        else // find correct sub-tree or insert new sub-tree
+          {
+            std::vector<std::string> remaining_id = erase_first(ids);
+
+            bool found = false;
+            for (auto it = sub_trees.begin(); it != sub_trees.end(); ++it)
+              {
+                // find out where to insert item
+                if ((*it)->id == remaining_id[0])
+                  {
+                    found = true;
+
+                    (*it)->insert(remaining_id, wall_time);
+                  }
+              }
+
+            if (found == false)
+              {
+                std::shared_ptr<TimerTree> new_tree;
+                new_tree.reset(new TimerTree());
+                new_tree->insert(remaining_id, wall_time);
+                sub_trees.push_back(new_tree);
+              }
+          }
+      }
+    else // the provided name does not fit to this tree
+      {
+        AssertThrow(false,
+                    ExcMessage("The name provided is " + ids[0] +
```

Comment:
Here and in the following line (and maybe elsewhere in the patch): since `ids[0]` can be any sort of text, the resulting error message can be difficult to read if, for example, someone provides `ids[0]="difficult to read"`. A better way is to make clear what is text and what is a quoted variable name:
```suggestion
                    ExcMessage("The name provided is <" + ids[0] + ">" +
```

Reply:
improved asserts here https://github.com/dealii/dealii/pull/11939/commits/c7dd464584f2cd84d71cd857684b5cfb755daf5a